### ISSUE TITLE
provide means to make packages relocatable

### ISSUE THREAD

**Issue by emmenlau** [conan-io/conan#5664]
- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

In my eyes relocation is one of the biggest hurdles to using package managers with C/C++ binaries. I have a humble request that conan should help to make packages relocatable. Specifically, conan should integrate methods that help when a package install path is different on the target host from the path on the source host, and the package contains hard coded path references.

This problem appears for a significant number of libraries on most Un*x and sometimes on Windows platforms. While at first it may seem better to solve the issue at the root (upstream libraries) in practice this is virtually impossible due to the sheer number of affected build systems and libraries.

The user story is: Many C++ libraries contain references to their install path. For example, high impact libraries Qt and HDF5 have binary and text configs with hard coded install paths. Installing these packages into a different path breaks the libraries. The same is true for a major number of pkg-config-enabled libraries as well as many autotools-built libraries. Furthermore, on Un*x, many libraries feature a fixed `RPATH` or `RUNPATH` that points to the install location. This effectively makes it very challenging for users for build libraries with path `X` and later install into path `Y`. However in multi-user environments or distributed build environments exactly this is often required, that install machines can not mimic the paths of CI machines.

In my humble opinion this issue can be best solved in conan. In practice the issue can *not* easily be solved at the source. Solving this in conan will have many benefits to the larger community:
 - Relocatable binary packages allow for much more flexibility in the build system and install system because paths do not need to match.
 - I.e. in cloud-based builds there can be tight constraints about paths that can be cumbersome to match on the target host
 - Relocatable binary packages will make conan significantly more useful because it enables new user groups.
 - Resolving this issue *without* the help of a package manager is cumbersome because it affects the gros of C/C++ libraries and it can be challenging because it may require knowledge of the install path on the target machine (that is often unknown at build time).

conan has all required information to aid in relocating packages:
- conan knows the package path at build time and can store it in the package
- conan knows the package install path on the target machine and can put it in place of the build time path
- conan already supports the concept of post-package- and post-install-hooks where this functionality can be integrated.

Therefore I'd like to suggest a "best effort" approach. The responsibility of creating actual post-package- and post-install-scripts that fixup paths should be with the package maintainer. Community will need to ensure packages are relocatable as required. conan should provide high-level methods that help in making paths relocatable. The two predominant types of paths are:
 - shared library `RPATH`/`RUNPATH` on platforms that support this concept
   - These can be modified using tools like `chrpath`
 - text path strings in configuration files
   - These can be modified by textual search-and-replace

To help in the process, the path handling methods may need to provide means to ensure that all types of paths are treated correctly. I.e. the path `/a/b/c` may or may not be the same as `//a\b\\\\c`.

**Comment by memsharded** [comment]
Hi @emmenlau 

We totally agree with this, and it is something we always have in mind with designing and implementing Conan. There are a number of things Conan already provides to help with this:

- For example, by default the CMake helpers remove RPATH information: https://docs.conan.io/en/latest/howtos/manage_shared_libraries/rpaths.html
- For cmake generated files, absolute paths are replaced with variables in https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#patch-config-paths-experimental

Furthermore, in general the important piece is that Conan packages contain very little embedded static configuration, but all the necessary information necessary for build systems (which sometimes need to be absolute paths) is generated at install time by Conan, but they are not inside the packages.

**Comment by emmenlau** [comment]
Dear @memsharded , thanks, this sounds very good. For us this feature is essential to adopt conan, and we're very much looking forward to this.

**Comment by memsharded** [comment]
Hi @emmenlau 

Great! Is there anything else specific that should be addressed regarding this issue? If it was just a general question, maybe we could close it now, as I think you are good to start moving forward as by default packages will be relocatable. New issues can be created anytime to discuss particular, more specific possible package relocation issues, looking forward to hearing from you. Thanks!

**Comment by emmenlau** [comment]
Hi @memsharded, I'm sorry that my previous post was not very clear. I think what you mentioned about existing functionality sounds like a good start. But the core part of the issue would be the means to search and replace the path strings in packages, in a robust way. I think conan has some of the prerequisites that are needed, but the complete solution does not exist.

Very concretely, I would like to have something like this:
 - The package install path should be stored somewhere in the package (in the manifest or similar) at packaging time, possibly by the post-package-hook
 - In the post-install-hook, I would like to add a script that can search and replace this aforementioned path with the actual install path on the target system.
   - To help implement this post-install-script, conan should implement search-and-replace-methods that can be conveniently called for this purpose of replacing paths in the target installation.

Does that make sense?

**Comment by emmenlau** [comment]
To be even more concrete, here is pseudo code that should illustrate the kind of post-install-script I think is required to make packages relocatable. It's implemented for an imaginary library called `mylib`. This imaginary library uses fixed `RPATH`, provides a `libmylib.la` file for helping the linker, and provides a pkg-config script as well:

```
post_install() {
    var original_install_path = conan.package.manifest.original_install_path;
    var target_install_path = conan.current.target_install_path;

    conan.tools.replace_rpath(original_install_path, target_install_path,
                    target_install_path + "/lib/libmylib.so");
    conan.tools.replace_text_path(original_install_path, target_install_path,
                    target_install_path + "/lib/libmylib.la");
    conan.tools.replace_text_path(original_install_path, target_install_path,
                    target_install_path + "/lib/kgconfig/mylib.pc");    
}
```

If conan could help in making this information `original_install_path`, `target_install_path` available and could provide high-level methods for replacing textual paths as well as rpaths, then package relocation becomes reasonably simple to be implemented on a broad scale!

**Comment by memsharded** [comment]
Hi @emmenlau 

> The package install path should be stored somewhere in the package (in the manifest or similar) at packaging time, possibly by the post-package-hook

I am not sure I understand this correctly. Storing the package install path inside the package itself is not necessary, as you cannot locate that package without the folder anyway. So Conan packages do not store any absolute path in it, including the package to itself.

**Comment by emmenlau** [comment]
Dear @memsharded , last time we spoke about this you mentioned that work is underway that conan would support relocating binary packages.

To make a concrete example, when building Qt into a target directory `/data/jenkins/binaries` then users would be able to install the conan package into `/home/builduser/binaries`. The relevant part is that conan (or the package install script) should help to change hardcoded references of the build-system path `/data/jenkins/binaries` (like they appear for example in library `*.la` files or often in pkg-config `*.pc` files) to the install path `/home/builduser/binaries`.

Is there progress on something like this?

**Comment by emmenlau** [comment]
Anyone?

**Comment by memsharded** [comment]
Hi @emmenlau 

Unfortunately, it has not been possible to progress here. We have some ideas, like having an ``install()`` step in recipes that allows to define this kind of logic, as well as targeting other folders rather than the cache. 
But other higher priorities were pushed much harder (package revisions and lockfiles, as they were very important for Continuous Integration, and ConanCenter new build infrastructure, because contributing to ConanCenter was difficult and ConanCenter was impossible to maintain). This is probably something that we will try to design for Conan 2.0 later this year.

What we got in the meantime is the ``-g deploy`` generator, that helps copying things out of the cache. Then, still the paths thing remains, in the meantime, I can only thing of a external script as a workaround to do the job. Maybe a custom ``deploy`` generator?

**Comment by blejdfist** [comment]
I need something similar when packaging a cross compilation SDK (built with buildroot).

The resulting files has references to the original build directory and has to be relocated after being moved. Buildroot even supplies a script that does this, so all I would have to do is run it after the package is installed.

Being able to execute a script inside the package install folder to fixup the paths would solve the problem for me.

**Comment by emmenlau** [comment]
I think this is really a blocker for a number of scenarios. Therefore it should be addressed in a clean and stable way. Running a post-install-script will go quite some way, but it will not be long before you also need to know the build path or the build-time temporary install path (pre-packaging).

For what its worth, conda has invested quite a lot to make this possible, they support full relocation for both text-strings as well as rpath since a long time.

Personally I prefer conan over conda so I will just wait patiently for it to be ready for my use case... :-) :-)

**Comment by OznOg** [comment]
I have similar issue, and using the relative RPATH mentionned in https://github.com/conan-io/conan/issues/1303 allowed to have binaries to find their library on any .conan tree. The same concern exists for configuration file and/or resources (things in /etc and /usr/share etc...) that needs to be accessed with relative path. This makes configurations hard to access from another conan package. For example, I use a server and conan packages for its plugins. But plugins need to "register" their name in the server config tree on install. As this install directory is not a "singleton" anymore, configs needs to be merge "by another mean".
As far as I know, on windows, custom data directories for applications also exists and I don't really understand how to have it work on any platform (as having different drives on windows leads to path that cannot be relative to each other (say binaries are on c: and data on d: ).
I'm still looking for some scalable solution.
Is there a conan "recommended way" for managing all resources files access?

**Comment by blackliner** [comment]
@emmenlau @memsharded 
I am looking forward to a first class support of RPATH modification by conan. Are there any news regarding this?

Currently I have to do this in my `def imports(self):` to be able to have a relocatable package:
```python
file_content = """file(RPATH_CHANGE FILE "${SOURCE_FILE}" OLD_RPATH "${OLD_RPATH}" NEW_RPATH "$ORIGIN")"""
tools.save("runpath.cmake", file_content)

for file in glob.glob("install/lib/*.so"):
    current_rpath = subprocess.check_output(["readelf", "-d", file]).decode("utf-8")
    match = re.search("Library runpath: \[(.*)\]", current_rpath)
    if match:
        self.run(
            [
                "cmake",
                f"-DOLD_RPATH={match.group(1)}",
                f"-DSOURCE_FILE={file}",
                "-P",
                "runpath.cmake",
            ]
        )
```

**Comment by emmenlau** [comment]
@blackliner For me this was a killer restriction in conan that it does not help with relocation. Changing the `RPATH` is one (important) aspect, but there are other aspects too. For example, `pkg-config` tends to hard-code install paths in its configs, and sometimes `cmake` configs do that too. Autotools often uses hard-coded install paths in its `.la` files. And Qt (at least until recently) hard-coded the install path inside the binaries.

But while the sources of this problem are many-fold, the solution is relatively manageable. Most of this can be solved in something like a week's work. But in my humble opinion, this should tightly integrate with conan's core functionality. An example workflow would be:
 - Recognize install paths in the (temporary) install location,
 - Replace them with specific (unique) tokens,
 - Replace the tokens at the package post-install time with the users install path
   - including replacement of `RPATH` on Un*x

All of this is neither hard nor complex. I can say so, because I implemented all of this in bash (for Windows, MacOS and several flavors of Linux), since conan did not have it.

But it is a certain effort to do these steps in a clean and concise manner. I do not think users should be required to "patch" this into conan's hooks. This is an essential requirement of software deployment and _needs_ to be addressed by means at the core of the software.

**Comment by solvingj** [comment]
@emmenlau What exactly did you do on Windows?

**Comment by emmenlau** [comment]
> @emmenlau What exactly did you do on Windows?

I'm sorry but I don't understand. Can you rephrase the question?

**Comment by solvingj** [comment]
>  I implemented all of this in bash (for Windows, MacOS and several flavors of Linux)

Since PE binaries don't have a concept of `RPATH`, I was wondering if you did something with Application Manifests:
https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests

**Comment by emmenlau** [comment]
> > I implemented all of this in bash (for Windows, MacOS and several flavors of Linux)
> 
> Since PE binaries don't have a concept of `RPATH`, I was wondering if you did something with Application Manifests:
> https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests

I see! That's an interesting question. I did not consider this. But I may look at it actually.

What I did are only two things:
 - Check if there are hard-coded references to the install path with textual and binary grep. The textual grep can be used with search and replace later to fix the path. And binary grep is used as a "safety guard" to error out, and kindly ask the user to change the upstream build (if possible :-) ).
 - Ensure validity of common install locations, like `dll` that should go to `bin` instead of `lib`. Many projects allow to override the detailed install location from the outside, i.e. with a cmake setting. But some projects hard-code a certain behavior. This makes it harder to run executables directly from the install path.

**Comment by gabyx** [comment]
@blackliner I think that kind of thing would be nice to have, but first of all if one uses CMake, the whole Cmake install for your Conan package A should use RPATHs e.g. 

Package A/
  - bin
    - `exe` (needs A.so)
  - lib/
    - `A.so` (needs B.so)
    - `B.so` (needs `fmt.so`, `libstdc++.so`)
  - external/lib
    - `fmt.so`
    - `libstdc++.so`
  
So in the above case using the following RPATHS in `exe`
- `$ORIGIN/../lib` 
- `$ORIGIN/../external/lib`  
can provide some solution.

But am I right, the solution for making `exe` runnable is to have conan package A (only the `exe`) with the appropriate dependencies,
and Users who need A, simply do a `conan deploy` and the `exe` can then be run without any problem??

**Comment by blackliner** [comment]
@gabyx agree, using `$ORIGIN` based `RPATH`s is the best way to do it. Unfortunately, not all open-source library give too much about correct CMake 🙁 

To really have a relocatable package, the customer might not want to use any conan commands at all. I come from the automotive/embedded world, where we can't expect our customers to install and run conan. We need to provide ready-to-eat packages, with everything needed inside. For this, we currently use a combination of the `deploy` generator, manual maintained `private=self.options.shared` requirements, deleting, copying and renaming files, all the way to modifying the files generated by the `cmake_find_package` generator. Very. Ugly. :-(

**Comment by emmenlau** [comment]
The issue with RPATH is indeed relevant. However I found it moderately easy to solve in almost all build systems that I tried (at least automake and cmake). Therefore I want to stress again that this issue is for the bigger picture, including but not limited to library search paths.

It is, for example, currently not possible to adjust absolute library directories of `pkgconfig` files at install time, if the install path differs from the build path. A significant list of packages creates `pkgconfig` files with absolute paths, sometimes not using pkgconfig itself, but i.e. with simple shell scripts or other tools. Such paths can lead to bad build problems if not adjusted, i.e. if the path should be actually available on the machine but contains a different version of the package! This happens i.e. if stable vs development versions are built and used in different conan roots.

But again, this is not limited to `pkgconfig` files. The same is relevant for Qt (See i.e. https://answers.informer.com/132035/patching-binaries-after-relocation-windows-qtbinpatcher), for some autotools files (typically `.la` files) and other configurations.

More generically, conan should check if a package contains the install path in any of its config files. If this is the case, it should at least issue a warning if such a package is later installed into a *different* target directory. Even better would be if such directories could be "adjusted" to point to the actual package install target directory.

**Comment by gabyx** [comment]
Regarding the RPATH troubles:

as @memsharded already stated, Conan does not enforce how CMake (as an example) is used. So any library which creates a cmake install folder which is not selfcontaining and using wrong (absolute paths etc...) RPATHs in its executables and libraries (e.g. also the missing `@rpath` for install names in `.dylib` files on macOS poses problems:
#10253) is entierly the problem of the package maintainer. The created Conan package then just imposes significant usage problems for consumers when they want to make their package relocatable (e.g. messing with RPATHs of dependencies etc...)

**Example:**
This comes to light, when for example on **macOS** a consumer uses the `fmt` library from the ConanCenter for its executable `exe`. Fmt is built by Conan with **naked install names** (no `@rpath/` prepended before `fmt.dylib`) which makes the consumer able to build the executable (e.g. dyn. linking `fmt`) but not able to run it in a self-contained install folder `installDir`. 

Why?: The external dependency `installDir/lib/fmt.dylib` won't be found by the loader in `installDir/bin/exe`:
The executable `installDir/bin/exe` might have (good practice IMO)  an `RPATH` `@executable_path/../lib` embedded, but it does not help because the loader does not consider the RPATH mechanism since the library to load is not named `@rpath/fmt.dylib` but `fmt.dylib`. The only way is to mess with some env variables `LD_LIBRARY_PATH`, which in turn is in contradiction to a self-containing `installDir`.

So @emmenlau's request to work more towards self-containing packages would help a lot. Also you can really be sure that when you install an end-user package (tarred self-contained install folder) which contains for example a server executable (service etc.) the whole package is self-contained. It runs out of the box, anywhere etc. so you could just untar the package (`installDir`) anywhere and it works... 

So far, as I understood, when working with Conan **to package the end-user artefacts** (the `exe` ) we solve this by 
- simply packaging only the `exe`, 
- specifying the dependencies inside the accompanying `conanfile.py`
- At the install site: with `conan install` 
  - letting Conan resolve the dependencies (which is good) by installing them into the local cache
  - generating the run environment by https://docs.conan.io/en/latest/reference/build_helpers/run_environment.html
  - launching the `exe` by first sourcing the environment scripts.

The above Conan approach of course works, but Debugging setups for `exe` in  IDEs become quite more intricate etc,
and its not really self-contained, it needs Conan at the place/machine where you install your stuff.

@memsharded: Is my understanding correct? And could`conan deploy` help by installing everything into a self-contained folder (package and its dependencies) from which wen then can launch `exe`?

**Comment by OznOg** [comment]
You point here the fact that packaging is usually really tight to the system you target. Standard Linux use canonical library paths, Android app bring all libs they need and Windows a mix of both.
Packaging a software for a delivery is more than just putting some binaries in an archive, some things are so system dependent that you must write them specifically: for example, in rpm packages, you need to specify a category for the software you are delivering (game, editor, multimedia, etc...), and no conan package will ever do it (would be meaningless).
IMHO using conan to deliver "final" binaries to a customers is a bad idea as you won't be able to deliver packaged that fit *their* environment, but packages that "made your (developer) life easier". Ends up that conan could be fine a setting up a dev env and ease testing against many versions of a given library, but in the end, the integration in the environment of your customer will remain necessary and trying not to do it by aggregating things built with conan will just end up making your customer's life harder.

**Comment by gabyx** [comment]
Von meinem iPhone gesendet

**Comment by gabyx** [comment]
@OznOg Ok thanks for the input. I agree, however whats is then a good way of delivering a final package? Maybe offtopic, I thought of a way of making the install directory of cmake self-contained up to the point of system specific libraries, and only those. So if one does not use Conan + some post processing steps, one has at one point to assemble all dependencies (libs, assets etc) put it into a folder structure, modifying RPATH or modify startup scripts, additional pre install scripts and configs and finally end up with a package (maybe one for the package mgr for your linux) which can be deployed… Is there any good approach?

**Comment by OznOg** [comment]
conan has a generator 'virtualrunenv' that generated scipts that set LD_LIBRARY_PATH to prepend the conan libraries to the system's one. One could imagine creating an installer from the deploy(ed) libraries and using a similar virtualrunenv script to set environment and run the binary. This is a classical way for having standalone things on *nix. Handling dependencies on the system itself is not done (if you need a special libc version, you may have issues), but works in general.
With such an installer, the RPATH could simply be ignored ... but I don't think conan provides a turn key solution for this (but worth checking, I don't know all conan things....)

**Comment by emmenlau** [comment]
@OznOg The problem that @gabyx raises is more complex than just finding the libraries. It's essentially about all package files (like configs etc) but even further than that (like systemd integration of services, or logging, or other things).

But I think we're getting a bit side-tracked by considering system integration. If this should really be tackled, then maybe consider what other generic application bundles for Linux are doing (like AppImage, snap, flatpack, etc...)?

My original idea is really just doing the same as `make && make install` typically does, but with the extra effort to allow changing the target directory after the fact (at the time the conan package is deployed into a possibly *different* target directory). Conan currently blindly assumes that packages would support such relocation. But I found that in practice, many many packages do *not* accommodate for this! They are not relocatable due to:
 - RPATH issues (runtime library resolution)
 - Linker path issues (link time library resolution)
 - Configuration path issues
 - Other internal hard coded path issues (like Qt plugin and resource paths...)

Most of this can be fixed with a simple search and replace of the target directory. But conan does not currently do that!

**Comment by OznOg** [comment]
I agree with you that config is some more work: one could imagine that using the same "wrapper script" idea, we could add some config path environment variable that the software would read. For integration with third party frameworks (say systemd, as you stated), rpm/deb usually have scriptlets run when installing the software, so I guess this could also be achieved when unpacking the software bundle done with conan... and here we did not talk about security and user rights.
But this ends up re-developping a complte packager over the distro, which is really overkill and that will probably never be as good as the native one....

**Comment by gabyx** [comment]
@emmenlau The first two points, Linker path issues and RPATH issues I thought are basically the first two topics which should be addressed by Conan, which it does, partly. The others things I put into the category of packaging, and Conan is not a fullblown packager (as pointed out), and are probably good to leave out of this discussion. Not even sure how you would regex/replace binaries and paths in your output artefact which Conan packages...

But I can see the point of providing better RPATH handling and relocatability when an executable/lib is built with Conan dependencies or also packaged into a new Conan package...

**Comment by emmenlau** [comment]
@gabyx I do not agree here, I'm sorry. I'm slightly frustrated why this obvious problem is being ignored for so long. I'll try to make an illustrative example, please try to follow me on this. In my humble understanding, the following is a valid and common thing to do with Conan, correct?
 - Build a package X with Conan into some target directory `A`
 - Package the resulting installation with Conan
 - Upload to a server, download again on the same or different machine
 - Extract the Conan package into target directory `B`

Now, many many packages use the absolute path `A` for certain things. For example, automake packages may add so-called `.la`-files in which they keep some helpful variables like a linker path `A`. Other packages may create `pkgconfig` files that reference absolute path `A`. When a user installs such a package into `B`, and builds against this package, they may get illegal references to `A` in their builds. Even worse, in a multi-user environment, the build server and the development desktop computer may be the same machine (this is our use case). When we install Conan packages into path `B`, they may continue to reference the CI installation in path `A`!

Do you see how this is just wrong? Conan blindly assumes packages would be relocatable, but many packages are not. And this is not only an issue of `RPATH`.

**Comment by emmenlau** [comment]
Due to this shortcoming of Conan, I've implemented my own buildsystem completely from scratch. In this process I've found countless examples of such absolute paths `A` in packages. The most prominent example is Qt, that until recently contained absolute target directories *in the actual binaries*. But many other packages have such problems of relocation in some form.

`RPATH` is even the simpler to solve, because a number of workarounds exist.

I've been kindly trying to promote this issue with the Conan team for three years now. I must say its becoming slightly frustrating.

**Comment by gabyx** [comment]
@emmenlau Ok, thanks now I understand. However, isn't this the problem of the underlying build system used? (automake, pkgconfig, not familiar with them) in the sense, that they should provide a way of creating a relocatable install. Your suggestion to enable path regex/replacements built into Conan when packaging seems one solution, but is it at the right place. I know its frustrating, but it seems to me rather broken window patching and fighting symptoms. But I understand, maybe its the only feasible way. Why is C++ ecosystem so damm hard... that is really frustrating, at least Conan tries to align some nasty things...

I only worked with the new CMakeDeps generator which writes cmake config files, that works pretty well, and at the same time only having pure binaries in the package (no cmake stuff etc, no cmake generated config files on install which are relocatable afaik)

**Comment by emmenlau** [comment]
> However, isn't this the problem of the underlying build system used?

Well, this is a difficult question. I do share your sentiment here. And for many packages, I've worked with the authors to add better relocation support.

But in my personal opinion, its not entirely fair to put the blame on the packages. There was never a promise made that `make install` would create a relocatable installation, or am I on the wrong track here? Its good practice to make one's code relocatable. But then again, it's open source, and we should be thankful for what we get.

In my own work, the only solution was to either never create binary packages, or to ensure with search and replace that all references of target directories are replaced.

If you want to build your own opinion, maybe install a larger number of binary conan packages, then grep for the original build or target directories in the install tree. In my personal experience, you should find at least one such reference per 10 packages.

This can lead to bad problems if the original build or target directories are available on the same machine as the user is now developing, since they may host different versions of packages! We lost countless hours with such problems as they are very hard to isolate.

**Comment by emmenlau** [comment]
For reference, here are a number of examples:
 - https://github.com/BlueBrain/HighFive creates file `share/HighFive/CMake/HighFiveTargets.cmake` that uses absolute paths for `INTERFACE_INCLUDE_DIRECTORIES` and other absolute paths.
 - The https://lemon.cs.elte.hu/trac/lemon graph library creates cmake config files with absolute paths.
 - https://github.com/catchorg/Catch2 creates `pkgconfig` files with absolute paths in `share/pkgconfig/catch2-with-main.pc`
 - https://github.com/postgres/postgres keeps the original build configuration in config file `include/server/pg_config.h`, possibly so that plugins and extensions can use it. This can contain absolute include and link paths.
 - https://github.com/glennrp/libpng has `bin/libpng16-config` that contains absolute paths
 - https://icu.unicode.org/ has `icu/69.1/pkgdata.inc` with absolute paths

These are just the random first examples I found, there are countless more in my build tree.

**Comment by emmenlau** [comment]
Note that some of these actually use `cmake` for their build! They may still provide `pkgconfig` configuration additionally to their cmake configuration, for example for backwards compatibility with `automake`. To make things worse, downstream cmake may actually use the pkgconfig configuration to detect those packages! So even if upstream and downstream both use relocatable cmake, the problem of absolute paths can still arise.

The only "guarantee" for users that packages are correctly relocated is to use search for the target path.

At the very least, conan should perform such a search and error out on creating the package, warning the user that it may not be fully relocatable. Even better would be a mechanism where conan "fixes" the paths magically.

**Comment by emmenlau** [comment]
@memsharded , any chance to consider this issue?

**Comment by dawidcha** [comment]
We have been successful in using hooks to do relocation operations for e.g. .la files, pkgconfig files, as well as other generated scripts so beloved of linux builds, like apxs for apache httpd.

There is a hook script which hooks `post_download_package`, and loads enough of the recipe for the package being downloaded to then call recipe.relocate(package_id)

The default relocate method can perform some common relocations, but can also be overridden for specific recipes to prevent default behaviour and/or add recipe-specific functionality.

We first ensure that the original absolute build path is written to the shared repo so we have a reference point. Then a couple of helper methods are provided to effect the relocations: for scripts and similar text files by simply doing a search/replace for that path with the absolute path of the local target repo in specified text files. RPATH relocations are done using patchelf.

Ultimately, it works well for us: if there is appetite, I'm very willing to see how possible it might be to contribute.

@emmenlau - I can share some ideas to see if they will work for you.
@memsharded - Is this something worth investigating? I know there have been a few threads on this topic, and I think some other plans for v2...

Default relocate method looks like: 
```
    def relocate(self, package_id):
        """
        Called by a hook that runs after this module has been downloaded from the shared global repo
        The default implementation relocates executable scripts found in 'bin' by convention and
        pkgconfig files found in lib/pkgconfig by convention 
        """
        self.relocate_text_files(glob.glob(os.path.join(self.package_folder, "bin", "*")))
        self.relocate_text_files(glob.glob(os.path.join(self.package_folder, "lib", "pkgconfig", "*")))
        self.relocate_text_files(glob.glob(os.path.join(self.package_folder, "lib", "*.la")))

        self.relocate_rpath(glob.glob(os.path.join(self.package_folder, "bin", "*")))
        self.relocate_rpath(glob.glob(os.path.join(self.package_folder, "lib", "*.so*")))   
```

**Comment by emmenlau** [comment]
Dear @dawidcha thanks a lot for this awesome head start! I think this is an absolutely crucial and underestimated feature. We currently do not use conan and I don't have resources to look into this. But it sounds great!

**Comment by gabyx** [comment]
@dawidcha: It would be interessting to see what you do in `self.relocate_rpath` ? 
We do some rpath (`runpath`) modifications when we `imports()` all libs into the build folder to use `$ORIGIN` or `@loader_path` (macOS) to make them referencing each other =).

**Comment by dawidcha** [comment]
@gabyx: The solution involves using the patchelf command to replace the runpath/rpath in any ELF binaries. To run pathchelf, we created a dependency in the recipe so its location can be retrieved from deps_cpp_info. Whether that's the best way of doing that is an open question, but it means we can write code something like this:
```
    BUILD_PACKAGE = ".prefix" # name of file containing the 'build prefix' code was added to write this at the end of the package operation

    def get_original_package_folder(self):
        if self.original_package_folder is None:
            original_loc_file = os.path.join(self.package_folder, self.BUILD_PACKAGE)
            if not os.path.lexists(original_loc_file):
                self.original_package_folder = ''
            else:
                with open(original_loc_file) as f:
                    ret = f.readline()
                 
                self.original_package_folder = '' if not ret else ret.strip()

        return self.original_package_folder

    def relocate_rpath(self, file_list):
        if not self.get_original_package_folder() or not sys.platform == "linux":
            return

        patchelf = os.path.join(self.deps_cpp_info["patchelf"].bin_paths[0], "patchelf")

        for fn in file_list:
            # process_output runs the given command and returns the output lines in an array
            file_type_info = process_output(["file", "-b", fn])
            if not (file_type_info and file_type_info[0].startswith("ELF ")):
                continue

            print("Relocating ELF file %s" % fn)
            rpath = process_output(["bash", "-c", "objdump -x %s | grep RPATH | sed \"s/^.* //\"" % fn])
            runpath = process_output(["bash", "-c", "objdump -x %s | grep RUNPATH | sed \"s/^.* //\"" % fn])

            if rpath:
                new_rpath = rpath[0].replace(self.get_data_folder(self.get_original_package_folder()), self.get_data_folder(self.package_folder))
                print("Changing RPATH from\n%s\nto\n%s" % (rpath[0], new_rpath))
                self.run("%s --force-rpath --set-rpath %s %s" % (patchelf, new_rpath, fn))
            
            if runpath:
                new_runpath = runpath[0].replace(self.get_data_folder(self.get_original_package_folder()), self.get_data_folder(self.package_folder))
                print("Changing RUNPATH from\n%s\nto\n%s" % (runpath[0], new_runpath))
                self.run("%s --set-rpath %s %s" % (patchelf, new_runpath, fn))
```

**Comment by memsharded** [comment]
I wanted to update here that Conan recently added support for the ``finalize()`` method, a way that package recipes can define logic that customize packages to their install location, see https://blog.conan.io/2024/09/03/New-finalize-method.html

Please have a look, it might be a good place to do some of the logic discussed in this thread. This is intended for installing packages inside the cache, not for ``deployers`` copying things outside of the cache.

I have re-read the whole thread again, and I think the topic is huge, there are different tools, build systems, platforms, etc, and still a bit outside of our scope and priorities. While I understand it is important in some cases, and we would like to be able to support this better, we have also been in continuous and growing engagements with tons of users, and it seems that there are still some many other higher priorities, which keep us constantly overwhelmed, so it is never possible to find time for this.

So in my experience, the best way to move this forward would be by leveraging users experience like yours, and work with relatively focused contributions. My proposal is:

- This ticket is not actionable, as it is too broad, it is the kind of ticket that can never be closed, even with huge efforts. This is bad for motivation, engagement and progress. So I suggest closing the ticket and follow up in new, focused ones
- Lets work in smaller, very focused batches. For example, if anyone is interested in contributing a tool for managing the RPATHs in Linux, then please open a new ticket with that very specific proposal. Propose an interface/usage example
- As long as we can think of an opt-in, low risk and extensible interface for this kind of tools, it is possible to move forward step by step, gather feedback from other users. We might want to propose a new ``conan.tools.relocate`` sub-package of tools, for example.
- We need to discuss and agree of the very specifics of the relocation, because those tools might be needed at ``finalize()`` time, or at ``deployer`` time, and the interfaces, importing, etc, needs to be prepared for this.

**Comment by memsharded** [comment]
Closing the ticket according to the previous comment: 

> This ticket is not actionable, as it is too broad, it is the kind of ticket that can never be closed, even with huge efforts. This is bad for motivation, engagement and progress. So I suggest closing the ticket and follow up in new, focused ones

New tickets following the above recommendations will be very welcome.
